libs/common/src/main/java/org/opensearch/common/WildcardMatcher.java

```diff
@@ -0,0 +1,499 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Class used to create matchers that can be used to test for matching patterns.
+ *
+ * WildcardMatcher supports matching using regex patterns.
+ *
+ * WildcardMatcher as supports wildcard (*) for matching any pattern and (?) for matching any single character.
+ *
+ * @opensearch.api
+ */
+public abstract class WildcardMatcher implements Predicate<String> {
```

Comment:
Why do we need this class? We already have `WildcardExpressionResolver` , I believe this is the one which we should be using since it also deals with close indices, etc.

Reply:
@reta This class is not related to indices, its a general utility used for wildcard matching.

Essentially, this is its usage:

```
Wildarcard matcher = WildcardMatcher.from("foo*", "bar");

matcher.test("foo1"); // true
matcher.test("foo"); // true
matcher.test("bar"); // true
matcher.test("bar1"); // false
matcher.test("baz"); // false
```

Reply:
Its used to create a matcher from the list of system indices/system index patterns

Reply:
Its extensively used in the security plugin and I'm proposing bringing it to core to support this use case, but if you think there is an existing alternative to use within core then I will update the PR.

Reply:
I could use [Regex.simpleMatch](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/regex/Regex.java#L152-L159)